### PR TITLE
Client side search

### PR DESCRIPTION
**Pull request by tomhazledine**
### Overview

Adds a search bar to the AppHeader for all mosaic sites.
Cursor_and_Modes_of_operation

Content included in the search index is set via the plugin options. `title`, `content`, and `route` (the page's relative URL) are included by default, and extra fields can be added with the `keys` option (the standard generator adds `description` as an example).

#### Shows results in real-time when making a query, with a maximum of ten results shown at any time.
Screenshot 2023-02-10 at 10 05 55

#### Handles typos/stop-words/etc.
Screenshot 2023-02-10 at 10 05 28

#### Shows a "no results found" message when no results are found
Screenshot 2023-02-10 at 10 05 41

### Things to add in the next sprint

i.e. things not 100% covered by this PR

* The "popper" element that contains the search-results dropdown is currently manually positioned. The plan is to update this to use the `useFloatingUI` hook provided by Salt. This should ideally be a more generic `<Popper/>` component that we can use throughout the site, so that will be better handled in a separate PR.
* Search "relevancy" is controlled by `fuseOptions` object in `packages/site-components/src/SearchInput/searchUtils.ts`. The next update will allow customisation of these values from the plugin options definition. *These values have a huge impact on which results are shown, and will most likely need to be configured for every new site.*
* More keyboard events. Search results are keyboard-navigable by tab and selectable with enter and space , but it would be good to add support for arrow keys to provide a more list-like experience.

### Commits

1. **add search-index plugin to standard generator**
2. **withSearchIndex middleware** - plumbs the search index data into the AppHeaderProps
3. **Search components** - adds the UI components required for search:`<SearchInput/>` (from salt-ds) and `<Results/>` (custom for this feature). Also adds [Fuse.js](https://fusejs.io/) (the open-source matching engine used by the search components) to the site-components package.

### CODE REVIEW

File: packages/site-middleware/src/withSearchIndex.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+export { SearchIndex };
+
+/**
+ * Fetch search-data.json
+ *
+ * @param url the search data file URL
+ * @returns search data JSON
+ */
+async function fetchSearchData(url: string) {
+  const response = await fetch(url, {
```
[the remark this answers is not in the document]
Comment (tomhazledine) [reply packages/site-middleware/src/withSearchIndex.ts:22]: Refactored completely to support loading from snapshots (and S3 snapshots)

File: packages/store/src/useAppHeader.ts [hunk 2]
```diff
@@ -26,20 +27,23 @@ export type AppHeaderMenu = Array<AppHeaderMenuLinksItem | AppHeaderMenuLinkItem
...
   HeaderControlsProps?: {
     searchEndpoint: string;
+    searchIndex: SearchIndexSlice;
   };
   menu: AppHeaderMenu;
 }
 
 export function useAppHeader(): AppHeader | undefined {
   const appHeader = useStore(state => state.sharedConfig?.header);
+  const allState = useStore(state => state);
```
[the remark this answers is not in the document]
Comment (tomhazledine) [reply packages/store/src/useAppHeader.ts]: Great suggestion. Extracted into `useSearchIndex` hook. Toyed with adding a `useSharedConfig` hook too, but the parts of search that relied on shared config were redundant anyway (tied to the external search service which is being retired).

File: packages/store/src/__tests__/useSearchIndex.test.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    ],
+    route: 'second/test/page/route'
+  }
+];
+
+const state: Partial<SiteState> = {
+  searchIndex: index
+};
+
+describe('GIVEN the `useSearchIndex` hook', () => {
```
Comment (DavieReid) [inline packages/store/src/__tests__/useSearchIndex.test.ts:27]: Nice

**Review by DavieReid — APPROVED** [review]
Hook refactor is much nicer, thanks for making the changes.

**Pull request merged**